### ISSUE TITLE
Feature request: block wrapping functionality

### ISSUE THREAD

**Issue by ohjimijimijimi** [jadejs/jade#2127]
Hello everybody, 

Jade already has support for **prepend** and **append** functionality when a block is extended. 

In the same way having a **wrap** functionality I think can help Jade to cover more use cases. The following gist has a full semplified example which explains the idea: https://gist.github.com/ohjimijimijimi/72c41aeda35ec4e5f9d4. 

Just to summarize here:
Here it is how this functionality can be used:

Imaging the following template:
``` jade
// template.jade
html
  body
    block page-container
      block page-header
      block page-content
      block page-footer
```

``` jade
// template-wrapped.jade
extends ./template.jade

block wrap page-container // or just wrap page-container
  .awesome-page-wrapper
```

and the result:
``` html
<html>
  <body>
    <div class="awesome-page-wrapper">
      <!-- whatever is in the inner blocks -->
    </div>
  </body>
</html>

Maybe there is already a way to do it. That would be cool! 

Comments are welcome :)

**Comment by ohjimijimijimi** [comment]
Just discovered a way to achieve the same behavior.

``` jade
// template-wrapped.jade
extends ./template.jade

block page-container
  .awesome-page-wrapper
    block page-header
    block page-content
    block page-footer
```

Unless anyone is interested I think this can be close -.-

**Comment by ForbesLindesay** [comment]
You can't do exactly what you're requesting, as we need to provide something to indicate **where** the content should go.

I'll use the proposed new syntax here, as I think being explicit about declaration vs. replacement is useful.  I would suggest that instead of this, you could have something like:

```jade
// template-wrapped.jade
extends ./template.jade

@replace page-container
  .awesome-page-wrapper
    @before page-container
```

Where `@before page-container` gets replaced with whatever the content of `page-container` was before it was `@replace`d.

That way we would have a nice desuggaring for prepend/append.

```jade
@prepend block-name
  div foo
```

desuggars to:

```jade
@replace block-name
  div foo
  @before block-name
```

```jade
@append block-name
  div foo
```

becomes

```jade
@replace block-name
  @before block-name
  div foo
```

The name `@before` probably needs some bike-shedding.  I was thinking `@default` but that's a little bit similar to `default` in `case`/`when`.  We could have something like `@parent` or `@parent-content`. I actually quite like `@yield` as something that mimics ES2015 generators and jade's own (soon to be deprecated) `yield` keyword in `include`s.  Another thought is that it could figure out the `block-name` from the `@replace` call higher up the stack, so you could just do:

```jade
// template-wrapped.jade
extends ./template.jade

@replace page-container
  .awesome-page-wrapper
    @before
```

**Comment by ohjimijimijimi** [comment]
Yes, I like the approach, and expliciting the declaration make everything more clear. 

I spend some time looking at the new proposed syntax. Taking that into consideration I think this make perfect sense. 

> The name ```@before``` probably needs some bike-shedding

Between all the proposed I still prefer ```@before``` because it has a time connotation, and, as you said

> Where ```@before``` page-container gets replaced with whatever the content of page-container was before it was ```@replaced```.

Maybe ```@origin``` or ```@orig``` could then be it.

> Another thought is that it could figure out the block-name from the @replace call higher up the stack, so you could just do

I like it.

At last, what can happen if someone use another block name? Compilation error? Let's imagine:
``` jade
// page.jade
@block page
  @block menu
  @block content
```

``` jade
// menu-left.jade
extends ./page.jade

@replace page
  .float-left
    @before content
```

This confuse me. Maybe the ```@before``` (or whatever is decided) should always go without parameter, so it can only reference the same block

**Comment by ForbesLindesay** [comment]
You could have something like:

```jade
// layout.jade
.left-column
  @block left-column
    h1 Default Left Column
.right-column
  @block right-column
    h1 Default Right Column
```

```jade
extends ./layout.jade

@replace left-column
  @before right-column

@replace right-coumn
  @before left-coumn
```

which could effectively swap the two columns.  i.e. `@before` means "before this file began being evaluated" rather than "before the current `@replace`".  I'm not sure that makes much sense though.

**Comment by ohjimijimijimi** [comment]
Yes, It make sense. It's better to don't limit the functionality just because it can create strange results. 
It's better to teach how to use it instead. 

Thanks also for the explanation on the ```@before``` concept. 

I don't have a lot of spare time... but, I would like to try to help in some way. How can I do that?
